docs/concepts/serving/executor/add-endpoints.md

```diff
@@ -153,18 +149,73 @@ class MyExecutor(Executor):
 If there is no `request_schema` and `response_schema`, the type hint is used to infer the schema. If both exist, `request_schema`
 and `response_schema` will be used.
 
-(endpoint-arguments)=
-## Arguments
-All Executor methods decorated by `@requests` need to follow the signature below to be usable as a microservice to be orchestrated either using {class}`~jina.Flow` or {class}`~jina.Deployment`.
 
-The `async` definition is optional.
+```{admonition} Note
+:class: note
 
-The endpoint signature looks like the following:
+When no type annotation or argument is provided, Jina assumes that [LegacyDocument](https://docs.docarray.org/API_reference/documents/documents/#docarray.documents.legacy.LegacyDocument) is the type used.
+This is intended to ease the transition from using Jina with `docarray<0.30.0` to using with the newer versions.
+```
 
-```python
-from typing import Dict, Union, Optional, TypeVar
+(executor-api)=
+## Executor API
+
+In this section we are going to see what API these methods decorated by `@requests` need to have so that Jina can serve them using {class}`~jina.Deployment` or {class}`~jina.Flow`.
+
+The main job of an Executor is to process `Documents` that are sent via the network. Executors can work on these `Documents` one by one or in batches.
+
+This is determined by adding the argument `doc` if you want your Executor to work on one Document at a time or `docs` if you want to work on batches of Documents.
+
+These APIs and the type annotations related to them, also affect the way your {ref}`OpenAPI looks when deploying the Executor <openapi-deployment>` with {class}`jina.Deployment` or {class}`jina.Flow` with HTTP protocol.
+
+(singleton-document)=
+### Single Document
+
+When using `doc` as a keyword argument, you need to add single `BaseDoc` as your request and response schema as seen in {ref}`the document type binding section <document-type-binding>`.
+
+Jina will make sure that even if multiple `Documents` are sent from the client, the Executor will process only one by one. 
+
+```{code-block} python
+---
+emphasize-lines: 13
+---
+from typing import Dict, Union, TypeVar
+from jina import Executor, requests
+from docarray import DocList, BaseDoc
+from pydantic import BaseModel
+
+T_input = TypeVar('T_input', bound='BaseDoc')
+T_output = TypeVar('T_output', bound='BaseDoc')
+
+class MyExecutor(Executor):
+    @requests
+    async def foo(
+        self,
+        doc: T_input,
+        **kwargs, 
+    ) -> Union[T_output, Dict, None]:
+        pass
+```
+
+Working on single Documents instead of in batches can make your interface clearer and your code cleaner. In many cases, like in Generative AI input does not usually come in batches
+and models can be heavy enough that they cannot profit from processing multiple inputs at the same time.
```

Comment:
```suggestion
Working on single Documents instead of  batches can make your interface and code cleaner. In many cases, like in Generative AI, input rarely comes in batches,
and models can be heavy enough that they cannot profit from processing multiple inputs at the same time.
```